### PR TITLE
samples: add Coded PHY peripheral and central

### PR DESCRIPTION
**Pull request by ChristianHirsch**
The samples are based on the heart-monitor samples
except that they are using Coded PHY for communication

Signed-off-by: Christian Hirsch <christian.hirsch@tuwien.ac.at>

### CODE REVIEW

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (ChristianHirsch) [reply samples/bluetooth/central_hr_coded/src/main.c:200]: I guess after registering the scan callback, scanning is started with `bt_le_scan_start(&scan_param, NULL);`, right?

Comment (joerchan) [reply samples/bluetooth/central_hr_coded/src/main.c:200]: Correct, forgot to include that.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (ChristianHirsch) [reply samples/bluetooth/peripheral_hr_coded/src/main.c:143]: Well, this is how it is implemented in the Zephyr [`peripheral_hr`](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.3.0/samples/bluetooth/peripheral_hr/src/main.c) sample.
If you are fine with it, I will remove it..

Comment (joerchan) [reply samples/bluetooth/peripheral_hr_coded/src/main.c:143]: I see. Best thing would be to align it with the rest of the NCS samples, like this:
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/bluetooth/central_bas/src/main.c#L318-L366

Comment (ChristianHirsch) [reply samples/bluetooth/peripheral_hr_coded/src/main.c:143]: Alright, this is an example for a central. But do you also have an example for a peripheral?

But do you think it is a good idea to add this to the sample? I for myself like to have clean samples that show me one specific thing. In this example it is to create a Coded PHY connection. Adding 50+ lines of code increases the complexity of the sample without any benefit for the core idea. This may confuse people. But that's just my opinion :)

Comment (joerchan) [reply samples/bluetooth/peripheral_hr_coded/src/main.c:143]: It should be the same for peripheral, apart from who initiates it.

In your samples none of them initiates security, so these won't be used.
I don't think having pairing in this sample is needed, so feel free to exclude it for the sample. Makes sense to limit it to just connection setup on LE Coded PHY.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:43]: if you enable CONFIG_BT_USER_PHY_UPDATE, you can use bt_conn_get_info to get the current PHY of the connection, it should return LE Coded PHY at this point.

**Comment by ChristianHirsch** [comment]
@joerchan I think I resolved all your concerns.

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/central_hr_coded/src/main.c:220]: If get_info returns err then phy_info->tx_phy/rx_phy will not be valid.
moved this into an else of the above statement. Can move const struct bt_conn_le_phy_info *phy_info; declaration down as well then.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:58]: If get_info returns err then phy_info->tx_phy/rx_phy will not be valid.
moved this into an else of the above statement. Can move const struct bt_conn_le_phy_info *phy_info; declaration down as well then.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:93]: return err (ENOEXEC is for the shell).

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:101]: return err (ENOEXEC is for the shell).

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:110]: adv_index not used for anything other than printing it. Consider removing it.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (joerchan) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:115]: Since both of these are set to zero you don't need this param, consider calling  `bt_le_ext_adv_start(adv, NULL);` instead.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (ChristianHirsch) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:82]: I changed the initialization as well.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (ChristianHirsch) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:89]: @joerchan Did your changes, and also changed the initialization of `struct bt_le_adv_param param`

**Comment by ChristianHirsch** [comment]
@joerchan Anything else that needs to be changed?

**Comment by joerchan** [comment]
@ChristianHirsch Thank you.

We will need approval from @b-gent as well for the documentation before we can merge.

**Comment by grochu** [comment]
@ChristianHirsch thanks for the contribution! My team would like to have an eye on this PR as well, since we are maintaining most of the BLE samples in *sdk-nrf*. This may also be a subject for verification - we should consider including these samples in CI to enable automatic regression.

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (KAGA164) [inline samples/bluetooth/central_hr_coded/src/main.c:4]: I am not sure that the licence is valid here. Could you confirm @joerchan or @carlescufi ?

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (KAGA164) [inline samples/bluetooth/central_hr_coded/src/main.c:232]: Please use the `gatt_dm` module instead of the Host API to discovery. This sample will be simpler then.

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (KAGA164) [inline samples/bluetooth/central_hr_coded/src/main.c:185]: Please use the scanning module then you can set filter on devices.

File: samples/bluetooth/peripheral_hr_coded/src/main.c
Comment (KAGA164) [inline samples/bluetooth/peripheral_hr_coded/src/main.c:4]: I am not sure about licence like above

File: samples/bluetooth/central_hr_coded/CMakeLists.txt
Comment (KAGA164) [inline samples/bluetooth/central_hr_coded/CMakeLists.txt:10]: Is it needed here ?

File: samples/bluetooth/peripheral_hr_coded/CMakeLists.txt
Comment (KAGA164) [inline samples/bluetooth/peripheral_hr_coded/CMakeLists.txt:8]: There is only one source file. To be consistent with other samples, please change to
```
# NORDIC SDK APP START
target_sources(app PRIVATE
  src/main.c
)
# NORDIC SDK APP END
```

File: samples/bluetooth/central_hr_coded/CMakeLists.txt
Comment (KAGA164) [inline samples/bluetooth/central_hr_coded/CMakeLists.txt:7]: There is only one source file. To be consistent with other samples, please change to
```
# NORDIC SDK APP START
target_sources(app PRIVATE
  src/main.c
)
# NORDIC SDK APP END
```

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (grochu) [inline samples/bluetooth/central_hr_coded/src/main.c:45]: This code contains a partial implementation of a Heart Rate Client. The NCS approach is to extract such code into separate re-usable components, e.g.  subsys/bluetooth/services/hids_c.c. Have you considered implementing a Heart Rate Client - maybe in a separate PR?

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (joerchan) [reply samples/bluetooth/central_hr_coded/src/main.c:4]: The license will have to be changed

**Comment by joerchan** [comment]
@ChristianHirsch Please amend the commit

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (ChristianHirsch) [reply samples/bluetooth/central_hr_coded/src/main.c:4]: To what do I need to change the license?

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (ChristianHirsch) [reply samples/bluetooth/central_hr_coded/src/main.c]: Well, not yet because I'm just using it is a sample and I do not use Heart Rates for other projects. Alternatively, I could rewrite the code and make a BAS client out of it and use the existing BAS implementation.

**Comment by ChristianHirsch** [comment]
I fixed most of the implementation related topics: @KAGA164 @grochu, I'm using the scan module and the discovery manager to find devices and its characteristics now.
@joerchan I'm still not sure if the licensing is ok..
The doc is still open..

**Comment by b-gent** [comment]
> @b-gent I will take care of it today. Is it ok to make separate commits, or do you prefer to amend the commits and force push to have everything in a single commit?

Separate commits would be better so that the diff is clearly visible. Then, the commit(s) should be squashed before merging.

File: samples/bluetooth/central_hr_coded/src/main.c
Comment (grochu) [reply samples/bluetooth/central_hr_coded/src/main.c]: OK, let's leave the HR client as it is. It may be refactored in a separate PR at a later stage.

**Review by grochu — APPROVED** [review]
Looks good to me, thanks for the updates. Please just resolve the comments about the license.

**Comment by joerchan** [comment]
@ChristianHirsch Pushed an update with the License update and the Testing and Dependencies sections to the READMEs, along with some other smaller changes.

We have to use this PR after all due to the Contributor License signing.
Also removed your name from the copyright, since that should be covered by the CLA signing, and only the company that runs the project has the name there as I understood it.

Let me know if this is ok with you.

@b-gent Is the documentation OK now?

File: samples/bluetooth/central_hr_coded/README.rst
Comment (utzig) [inline samples/bluetooth/central_hr_coded/README.rst:18]: This block could be changed to:

```
.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340pdk_nrf5340_cpuapp_and_cpuappns, nrf52840dongle_nrf52840, nrf52840dk_nrf52840
```

And then you could discard the doc/nrf/includes/boardname_tables/sample_boardnames.txt changes from the PR.

**Review by utzig — CHANGES_REQUESTED** [review]

**Comment by b-gent** [comment]
> @ChristianHirsch Pushed an update with the License update and the Testing and Dependencies sections to the READMEs, along with some other smaller changes.
> 
> We have to use this PR after all due to the Contributor License signing.
> Also removed your name from the copyright, since that should be covered by the CLA signing, and only the company that runs the project has the name there as I understood it.
> 
> Let me know if this is ok with you.
> 
> @b-gent Is the documentation OK now?

I'll be able to check tomorrow

File: samples/bluetooth/central_hr_coded/README.rst
Comment (joerchan) [reply samples/bluetooth/central_hr_coded/README.rst:18]: @b-gent Since the table was added by you, do you agree with this?

**Comment by utzig** [comment]
I added a DNM label only while testing a CI feature, not a blocking really, and should be able to remove it soon...

File: samples/bluetooth/central_hr_coded/README.rst
Comment (b-gent) [reply samples/bluetooth/central_hr_coded/README.rst:18]: Yes, this is a very recent update to how we do these tables, I can change this tomorrow.

File: samples/bluetooth/peripheral_hr_coded/README.rst
Comment (utzig) [inline samples/bluetooth/peripheral_hr_coded/README.rst:39]: Documentation build is failing because the `central_hr_coded` label does not exist.

Comment (b-gent) [reply samples/bluetooth/peripheral_hr_coded/README.rst]: yes, this is fixed now

**Comment by b-gent** [comment]
Adding a commit with some further doc review. I added some content so please check it before squashing.

**Comment by ChristianHirsch** [comment]
@joerchan Yes, license update is fine! I added this, because I had no idea what to add. But now I know it for the next time.

@b-gent I will make separate commits the next time and squash it as you mentioned it. Sry for that!

File: samples/bluetooth/peripheral_hr_coded/README.rst
Comment (joerchan) [reply samples/bluetooth/peripheral_hr_coded/README.rst:26]: Fair assumption.

**Comment by joerchan** [comment]
Squashed commits:
@b-gent Only made one change, "with a strong enough signal" to "in range" since we don't check the signal strength.

**Review by b-gent — APPROVED** [review]

**Review by utzig — APPROVED** [review]
LGTM

**Review by KAGA164 — APPROVED** [review]

**Pull request merged**
